pkg/kubelet/cm/devicemanager/manager.go

```diff
@@ -350,20 +358,29 @@ func (m *ManagerImpl) isVersionCompatibleWithPlugin(versions []string) bool {
 	return false
 }
 
-func (m *ManagerImpl) allocatePodResources(pod *v1.Pod) error {
-	devicesToReuse := make(map[string]sets.String)
-	for _, container := range pod.Spec.InitContainers {
-		if err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
-			return err
+func (m *ManagerImpl) allocatePodResources(pod *v1.Pod, container *v1.Container) error {
+	if _, ok := m.devicesToReuse[string(pod.UID)]; !ok {
+		m.devicesToReuse[string(pod.UID)] = make(map[string]sets.String)
+	}
+	// If pod entries to m.devicesToReuse other than the current pod exist, delete them.
+	for podUID := range m.devicesToReuse {
+		if podUID != string(pod.UID) {
+			delete(m.devicesToReuse, podUID)
 		}
-		m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, devicesToReuse)
 	}
-	for _, container := range pod.Spec.Containers {
-		if err := m.allocateContainerResources(pod, &container, devicesToReuse); err != nil {
-			return err
+	for _, initContainer := range pod.Spec.InitContainers {
+		if container.Name == initContainer.Name {
+			if err := m.allocateContainerResources(pod, container, m.devicesToReuse[string(pod.UID)]); err != nil {
+				return err
+			}
+			m.podDevices.addContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
+			return nil
 		}
-		m.podDevices.removeContainerAllocatedResources(string(pod.UID), container.Name, devicesToReuse)
 	}
+	if err := m.allocateContainerResources(pod, container, m.devicesToReuse[string(pod.UID)]); err != nil {
+		return err
+	}
+	m.podDevices.removeContainerAllocatedResources(string(pod.UID), container.Name, m.devicesToReuse[string(pod.UID)])
```

Comment:
so if we were to keep error earlier, would this map grow unbounded?  i am just trying to make sure there isnt an unbounded growth of devices to reuse.

Reply:
I haven’t reviewed this in detail yet, but the idea was to simply take the local devicesToReuse variable in allocatePodResources and make it a type variable on the overall manager. Once this is done, the loop in the current allocatePodResources function should be unrolled and split across a loop in the TopologManager which Allocate() on each container. The main logic from allocatePodDevices should then be pushed directly into Allocate(). If any mutetest were necessary before, they would still be necessary now.

For reference, the removal happens asynchronous here https://github.com/kubernetes/kubernetes/blob/84dfce3fed075f83b8932776ca6009f60f92979a/pkg/kubelet/cm/devicemanager/manager.go#L658, lazily cleaning up the map for terminated pods, just before allocating devices to new pods.

Reply:
My takeaway from this discussion is that the only change I need make is to move logic from `allocatePodResources` back into `Allocate()` as it is now unnecessary to have the additional function.

Not mentioned here, but I will also remove the `TODO:` comment describing the new flow: https://github.com/kubernetes/kubernetes/blob/b5b395364cfdc4ea984d4f966a58fdd53edb82d5/pkg/kubelet/cm/devicemanager/manager.go#L390-L408 
I had just left this in for reviewing purposes.

Otherwise, I will leave the PR open for @klueska to give a final review

Reply:
@klueska thanks for the clarification.